File: frontend/src/api/cdrQuote.ts
/** CDR 智能报价 API 封装 */
import api from './index'
import type { PaginatedData, QuoteListResponse } from '@/types/api'

export type CDRQueryParams = Record<string, string | number | boolean | undefined>

export interface PricingTraceStep {
  rule_code: string
  description: string
  input_value?: Record<string, unknown>
  output_value?: Record<string, unknown>
}

export interface PricingResult {
  billable_quantity: string
  unit_price: string
  subtotal_amount: string
  material_cost: string
  process_cost: string
  startup_fee: string
  total_cost: string
  discount_amount: string
  tax_amount: string
  total_amount: string
  minimum_charge_applied: boolean
  requires_approval: boolean
  warnings: string[]
  pricing_trace: PricingTraceStep[]
}

export interface PricingRequest {
  product_id: string
  material_id?: string
  quantity?: number
  width_mm?: number
  height_mm?: number
  length_m?: number
  process_ids?: string[]
  customer_id?: string
  tax_rate?: number
}

export interface QuoteLineProcessInput {
  process_id: string
  billing_quantity?: number
  unit?: string
  unit_price?: number
}

export interface QuoteLineInput {
  product_id?: string
  material_id?: string
  item_name: string
  description?: string
  material_process?: string
  width?: number
  width_unit?: string
  height?: number
  height_unit?: string
  width_mm?: number
  height_mm?: number
  length_m?: number
  quantity?: number
  unit?: string
  use_area?: boolean
  pieces?: number
  unit_price?: number
  process_fee?: number
  installation_fee?: number
  design_fee?: number
  transport_fee?: number
  other_fee?: number
  remark?: string
  image_url?: string
  sort_order?: number
  group_name?: string
  manual_adjustment?: number
  manual_reason?: string
  processes?: QuoteLineProcessInput[]
}

export interface QuoteLineProcess {
  id: string
  process_id: string
  billing_quantity: string
  unit_price: string
  amount: string
}

export interface QuoteVersionLine {
  id: string
  line_no: number
  product_id?: string
  material_id?: string
  item_name: string
  description: string
  material_process?: string
  width?: string
  width_unit?: string
  height?: string
  height_unit?: string
  width_mm?: string
  height_mm?: string
  length_m?: string
  quantity: string
  unit?: string
  use_area?: boolean
  pieces?: string
  unit_price: string
  amount: string
  estimated_cost: string
  process_fee: string
  installation_fee: string
  design_fee: string
  transport_fee: string
  other_fee: string
  remark?: string
  image_url?: string
  sort_order: number
  group_name?: string
  source: string
  requires_approval: boolean
  processes: QuoteLineProcess[]
}

export interface QuoteVersion {
  id: string
  quote_id: string
  version_no: number
  status: string
  subtotal_amount: string
  total_amount: string
  estimated_cost: string
  estimated_profit: string
  estimated_margin: string
  notes?: string
  created_by?: string
  created_at?: string
  lines: QuoteVersionLine[]
}

export interface QuoteApproval {
  id: string
  quote_id: string
  quote_version_id?: string
  approval_type: string
  status: string
  reason?: string
  decision_comment?: string
  requested_by?: string
  approver_id?: string
  created_at?: string
  decided_at?: string
}

export interface PriceRuleInput {
  code: string
  name: string
  priority?: number
  conditions_json?: Record<string, unknown>
  actions_json?: Record<string, unknown>
  conflict_policy?: string
}

export interface PriceRuleSetInput {
  code: string
  name: string
  effective_from?: string
  effective_to?: string
  description?: string
  rules?: PriceRuleInput[]
}

export interface PriceRuleSet extends PriceRuleSetInput {
  id: string
  version: number
  status: string
}

export interface CustomerAgreementInput {
  customer_id: string
  product_id?: string
  material_id?: string
  process_id?: string
  pricing_method: string
  price_value: number
  minimum_charge?: number
  discount_rate?: number
  effective_from: string
  effective_to?: string
  remark?: string
}

export interface CustomerAgreement {
  id: string
  customer_id: string
  product_id?: string
  material_id?: string
  process_id?: string
  pricing_method: string
  price_value: string
  minimum_charge: string
  discount_rate: string
  effective_from: string
  effective_to?: string
  remark?: string
}

export interface CDRAuditLog {
  id: string
  actor_id?: string
  action: string
  reason?: string
  created_at?: string
}

export interface CDRQuote extends QuoteListResponse {
  doc_no?: string
  tax_rate?: number
  customer?: { name?: string }
}

export interface CDRQuoteCreate {
  project_name: string
  customer_id?: string
  customer_name?: string
  tax_rate?: number
  status?: string
}

export interface DesignAttachment {
  id: string
  filename: string
  file_size: number
  file_type: string
  created_at?: string
}

export interface ParsedShape {
  type: string
  width_mm: number
  height_mm: number
  area_m2: number
  quantity: number
  label: string
}

export interface SvgParseResult {
  document_width_mm?: number
  document_height_mm?: number
  shapes: ParsedShape[]
  shape_count: number
  total_area_m2?: number
  filename: string
  attachment_id: string
}

export interface CDRSuggestedLine {
  description?: string
  item_name?: string
  width_mm?: number
  height_mm?: number
  quantity?: number
  unit?: string
  material_suggestion?: string
}

export interface AiAssistResult {
  project_name: string
  files: string[]
  ai_suggestions:
    | CDRSuggestedLine[]
    | { items?: CDRSuggestedLine[]; lines?: CDRSuggestedLine[]; raw?: string }
}

/** 报价试算 */
export function calculatePricing(data: PricingRequest) {
  return api.post<PricingResult>('/cdr/pricing/calculate', data)
}

export function createCDRQuote(data: CDRQuoteCreate) {
  return api.post<CDRQuote>('/cdr/quotes', data)
}

/** 创建报价版本 */
export function createQuoteVersion(
  quoteId: string,
  data: { notes?: string; lines: QuoteLineInput[] },
) {
  return api.post<QuoteVersion>(`/cdr/quotes/${quoteId}/versions`, data)
}

export function getLatestVersion(quoteId: string) {
  return api.get<QuoteVersion | null>(`/cdr/quotes/${quoteId}/versions/latest`)
}

export function listVersions(quoteId: string) {
  return api.get<QuoteVersion[]>(`/cdr/quotes/${quoteId}/versions`)
}

export function listApprovals(quoteId: string) {
  return api.get<QuoteApproval[]>(`/cdr/quotes/${quoteId}/approvals`)
}

export function requestApproval(
  quoteId: string,
  data: { approval_type: string; reason?: string },
) {
  return api.post<QuoteApproval>(`/cdr/quotes/${quoteId}/approvals`, data)
}

export function approveQuote(approvalId: string, comment?: string) {
  return api.post<{ status: string }>(`/cdr/approvals/${approvalId}/approve`, { comment })
}

export function rejectQuote(approvalId: string, comment?: string) {
  return api.post<{ status: string }>(`/cdr/approvals/${approvalId}/reject`, { comment })
}

export function listRuleSets() {
  return api.get<PriceRuleSet[]>('/cdr/rule-sets')
}

export function createRuleSet(data: PriceRuleSetInput) {
  return api.post<Pick<PriceRuleSet, 'id' | 'code' | 'name'>>('/cdr/rule-sets', data)
}

export function listCustomerAgreements(customerId?: string) {
  const params = customerId ? { customer_id: customerId } : {}
  return api.get<CustomerAgreement[]>('/cdr/customer-agreements', { params })
}

export function createCustomerAgreement(data: CustomerAgreementInput) {
  return api.post<Pick<CustomerAgreement, 'id' | 'customer_id'>>('/cdr/customer-agreements', data)
}

export function listAuditLogs(quoteId: string) {
  return api.get<CDRAuditLog[]>(`/cdr/quotes/${quoteId}/audit-logs`)
}

export function deleteCDRQuote(quoteId: string) {
  return api.del<{ deleted: boolean }>(`/cdr/quotes/${quoteId}`)
}

export function convertToOrder(quoteId: string) {
  return api.post<{ id: string; doc_no: string }>(`/cdr/quotes/${quoteId}/convert-to-order`)
}

export function getCDRQuote(quoteId: string) {
  return api.get<CDRQuote>(`/cdr/quotes/${quoteId}`)
}

export function listCDRQuotes(params?: CDRQueryParams) {
  return api.get<PaginatedData<CDRQuote>>('/cdr/quotes', { params })
}

export function uploadDesignFile(quoteId: string, file: File) {
  const form = new FormData()
  form.append('file', file)
  return api.post<DesignAttachment>(`/cdr/quotes/${quoteId}/upload`, form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

export function listDesignAttachments(quoteId: string) {
  return api.get<DesignAttachment[]>(`/cdr/quotes/${quoteId}/attachments`)
}

export function deleteDesignAttachment(attId: string) {
  return api.del<{ deleted: boolean }>(`/cdr/attachments/${attId}`)
}

export function parseSvgAttachment(attId: string) {
  return api.post<SvgParseResult>(`/cdr/attachments/${attId}/parse-svg`)
}

export function aiAssistFromDescription(quoteId: string, description: string) {
  return api.post<AiAssistResult>(`/cdr/quotes/${quoteId}/ai-assist-description`, { description })
}

File: backend/app/models/cdr_quote.py
"""CDR 智能报价——新增领域模型。

设计原则：
- 复用现有 customers、business_documents(作为报价 header)、users 等表
- 报价版本/明细/审批/审计独立建表，不与现有 BusinessDocumentItem 耦合
- 金额全部使用 NUMERIC，禁止 float
- 每次报价快照保存规则版本、价格快照、图稿摘要
"""

import uuid
from decimal import Decimal
from datetime import date, datetime

from sqlalchemy import Boolean, Date, DateTime, Integer, Numeric, String, Text, ForeignKey, func
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, SoftDeleteMixin


# ── 1. 定价规则引擎 ─────────────────────────────────────────────


class PriceRuleSet(Base, TimestampMixin):
    """定价规则集——分组和版本控制。"""
    __tablename__ = "price_rule_sets"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    code: Mapped[str] = mapped_column(String(64), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    version: Mapped[int] = mapped_column(Integer, default=1)
    status: Mapped[str] = mapped_column(String(32), default="draft", comment="draft | published | archived")
    effective_from: Mapped[str | None] = mapped_column(String(32), nullable=True, comment="YYYY-MM-DD")
    effective_to: Mapped[str | None] = mapped_column(String(32), nullable=True, comment="YYYY-MM-DD")
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    published_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    published_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)

    rules: Mapped[list["CdrPriceRule"]] = relationship(back_populates="rule_set", lazy="selectin", cascade="all, delete-orphan")


class CdrPriceRule(Base, TimestampMixin):
    """定价规则——触发条件 + 计算动作。"""
    __tablename__ = "cdr_price_rules"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    rule_set_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("price_rule_sets.id"), nullable=False)
    code: Mapped[str] = mapped_column(String(64), nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    priority: Mapped[int] = mapped_column(Integer, default=0)
    conditions_json: Mapped[dict] = mapped_column(JSONB, default=dict, comment="触发条件列表")
    actions_json: Mapped[dict] = mapped_column(JSONB, default=dict, comment="计算动作列表")
    conflict_policy: Mapped[str] = mapped_column(String(32), default="higher_priority_wins", comment="冲突策略")
    active: Mapped[bool] = mapped_column(Boolean, default=True)

    rule_set: Mapped["PriceRuleSet"] = relationship(back_populates="rules")


# ── 2. 客户协议价 ───────────────────────────────────────────────


class CustomerPriceAgreement(Base, TimestampMixin):
    """客户专项价格协议。"""
    __tablename__ = "customer_price_agreements"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    customer_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("customers.id"), nullable=False)
    product_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=True)
    material_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("materials.id"), nullable=True)
    process_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("processes.id"), nullable=True)
    pricing_method: Mapped[str] = mapped_column(String(32), nullable=False, comment="area | length | quantity | fixed")
    price_value: Mapped[Decimal] = mapped_column(Numeric(14, 4), nullable=False)
    minimum_charge: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    discount_rate: Mapped[Decimal] = mapped_column(Numeric(8, 4), default=1.0, comment="折扣率：0.8=打8折")
    effective_from: Mapped[str] = mapped_column(String(32), nullable=False, comment="YYYY-MM-DD")
    effective_to: Mapped[str | None] = mapped_column(String(32), nullable=True, comment="YYYY-MM-DD")
    approved_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    remark: Mapped[str | None] = mapped_column(Text, nullable=True)


# ── 3. 报价核心 ─────────────────────────────────────────────────


class QuoteVersion(Base, TimestampMixin):
    """报价版本——每次保存生成新版本，历史版本不可修改。

    关联 quotes 作为报价 header，
    该表存储版本快照和明细。
    """
    __tablename__ = "quote_versions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    quote_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("business_documents.id"), nullable=False
    )
    version_no: Mapped[int] = mapped_column(Integer, nullable=False)
    pricing_rule_set_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("price_rule_sets.id"), nullable=True
    )
    status: Mapped[str] = mapped_column(String(32), default="draft", comment="draft | review | approved | rejected")

    # 金额汇总
    subtotal_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    discount_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    tax_rate: Mapped[Decimal] = mapped_column(Numeric(8, 4), default=0)
    tax_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    total_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    estimated_cost: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    estimated_profit: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    estimated_margin: Mapped[Decimal] = mapped_column(Numeric(8, 4), default=0)

    # 快照
    snapshot_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True, comment="报价快照（客户/产品/价格快照）")
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)

    # 关系
    lines: Mapped[list["QuoteLine"]] = relationship(back_populates="version", lazy="selectin", cascade="all, delete-orphan")


class QuoteLine(Base, TimestampMixin):
    """报价明细行。"""
    __tablename__ = "quote_lines"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    version_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("quote_versions.id"), nullable=False)
    line_no: Mapped[int] = mapped_column(Integer, nullable=False)

    product_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=True)
    material_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("materials.id"), nullable=True)
    description: Mapped[str] = mapped_column(String(500), nullable=False)
    material_process: Mapped[str | None] = mapped_column(String(500), nullable=True)

    # 几何/数量
    width: Mapped[Decimal | None] = mapped_column(Numeric(12, 3), nullable=True)
    width_unit: Mapped[str | None] = mapped_column(String(16), nullable=True)
    height: Mapped[Decimal | None] = mapped_column(Numeric(12, 3), nullable=True)
    height_unit: Mapped[str | None] = mapped_column(String(16), nullable=True)
    width_mm: Mapped[Decimal | None] = mapped_column(Numeric(12, 3), nullable=True)
    height_mm: Mapped[Decimal | None] = mapped_column(Numeric(12, 3), nullable=True)
    length_m: Mapped[Decimal | None] = mapped_column(Numeric(12, 3), nullable=True)
    quantity: Mapped[Decimal] = mapped_column(Numeric(14, 3), default=1)
    unit: Mapped[str | None] = mapped_column(String(32), nullable=True)
    use_area: Mapped[bool] = mapped_column(Boolean, default=False)
    pieces: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True, comment="件数")

    # 计费
    billable_quantity: Mapped[Decimal] = mapped_column(Numeric(14, 4), default=0, comment="计费数量（含损耗）")
    unit_price: Mapped[Decimal] = mapped_column(Numeric(14, 4), default=0)
    amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    estimated_cost: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    process_fee: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    installation_fee: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    design_fee: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    transport_fee: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    other_fee: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    remark: Mapped[str | None] = mapped_column(Text, nullable=True)
    image_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    group_name: Mapped[str | None] = mapped_column(String(255), nullable=True)

    # 手工调整
    manual_adjustment: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    manual_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    source: Mapped[str] = mapped_column(String(32), default="auto", comment="auto | manual")

    # 审批标记
    requires_approval: Mapped[bool] = mapped_column(Boolean, default=False)

    # 价格执行明细
    pricing_trace_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True, comment="规则执行过程")

    version: Mapped["QuoteVersion"] = relationship(back_populates="lines")
    processes: Mapped[list["QuoteLineProcess"]] = relationship(back_populates="line", lazy="selectin", cascade="all, delete-orphan")


class QuoteLineProcess(Base, TimestampMixin):
    """报价行工艺明细。"""
    __tablename__ = "quote_line_processes"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    line_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("quote_lines.id"), nullable=False)
    process_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("processes.id"), nullable=False)
    billing_quantity: Mapped[Decimal] = mapped_column(Numeric(14, 4), default=0)
    unit: Mapped[str | None] = mapped_column(String(32), nullable=True)
    unit_price: Mapped[Decimal] = mapped_column(Numeric(14, 4), default=0)
    amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    cost_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), default=0)
    pricing_trace_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    line: Mapped["QuoteLine"] = relationship(back_populates="processes")


# ── 4. 审批与审计 ───────────────────────────────────────────────


class QuoteApproval(Base, TimestampMixin):
    """报价审批记录。"""
    __tablename__ = "quote_approvals"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    quote_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("business_documents.id"), nullable=False)
    quote_version_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("quote_versions.id"), nullable=True)
    approval_type: Mapped[str] = mapped_column(String(32), nullable=False, comment="low_margin | over_discount | price_override | high_value")
    requested_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    approver_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    status: Mapped[str] = mapped_column(String(32), default="pending", comment="pending | approved | rejected")
    reason: Mapped[str | None] = mapped_column(Text, nullable=True, comment="请求原因")
    decision_comment: Mapped[str | None] = mapped_column(Text, nullable=True)
    decided_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)


class QuoteAuditLog(Base):
    """报价审计日志。"""
    __tablename__ = "quote_audit_logs"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    quote_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("business_documents.id"), nullable=False)
    quote_version_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("quote_versions.id"), nullable=True)
    actor_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    action: Mapped[str] = mapped_column(String(64), nullable=False)
    before_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    after_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    ip_address: Mapped[str | None] = mapped_column(String(64), nullable=True)
    device_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("cdr_devices.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())


# ── 5. CDR 设备与图稿采集 ────────────────────────────────────────


class CdrDevice(Base, TimestampMixin):
    """CDR 插件注册设备。"""
    __tablename__ = "cdr_devices"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    device_code: Mapped[str] = mapped_column(String(64), unique=True, nullable=False)
    device_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    employee_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    machine_fingerprint_hash: Mapped[str] = mapped_column(String(128), nullable=False)
    plugin_version: Mapped[str | None] = mapped_column(String(32), nullable=True)
    bridge_version: Mapped[str | None] = mapped_column(String(32), nullable=True)
    coreldraw_versions_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    last_seen_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    status: Mapped[str] = mapped_column(String(32), default="active", comment="active | revoked")
    revoked_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)


class CdrCaptureSession(Base, TimestampMixin):
    """图稿采集会话——CDR 插件每次"发送到 ERP"创建一条。"""
    __tablename__ = "cdr_capture_sessions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    session_code: Mapped[str] = mapped_column(String(64), unique=True, nullable=False)
    device_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("cdr_devices.id"), nullable=True)
    employee_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    document_name: Mapped[str | None] = mapped_column(String(500), nullable=True)
    document_path_hash: Mapped[str | None] = mapped_column(String(128), nullable=True)
    coreldraw_version: Mapped[str | None] = mapped_column(String(32), nullable=True)
    page_index: Mapped[int] = mapped_column(Integer, default=0)
    page_name: Mapped[str | None] = mapped_column(String(128), nullable=True)
    selection_count: Mapped[int] = mapped_column(Integer, default=0)
    drawing_fingerprint: Mapped[str | None] = mapped_column(String(128), nullable=True)
    capture_payload_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    warnings_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    captured_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    expires_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)


class DrawingSnapshot(Base, TimestampMixin):
    """图稿快照——报价确认/转订单时冻结的图稿摘要。"""
    __tablename__ = "drawing_snapshots"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    snapshot_code: Mapped[str] = mapped_column(String(64), unique=True, nullable=False)
    capture_session_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("cdr_capture_sessions.id"), nullable=True)
    quote_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("business_documents.id"), nullable=True)
    quote_version_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("quote_versions.id"), nullable=True)
    drawing_fingerprint: Mapped[str | None] = mapped_column(String(128), nullable=True)
    geometry_summary_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True, comment="几何摘要：宽高/面积/长度")
    object_summary_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True, comment="对象摘要：数量/类型分布")
    preview_file_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    created_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)

class QuoteGeometry(Base, TimestampMixin):
    """报价行几何分析——孔洞/重叠/曲线/板材套料数据。

    每个报价明细行一条记录，存储几何分析结果。
    is_estimated=True 表示该值是基于算法估算，非精确测量值。
    """
    __tablename__ = "quote_geometry"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    quote_line_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("quote_lines.id"), nullable=True, unique=True,
        comment="关联报价明细行"
    )
    quote_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("business_documents.id"), nullable=True,
        comment="关联报价 header（便于查询）"
    )

    # ── 孔洞与净面积 ──
    net_area_mm2: Mapped[Decimal | None] = mapped_column(Numeric(16, 3), nullable=True, comment="净面积（包围盒-孔洞）mm²")
    hole_area_mm2: Mapped[Decimal | None] = mapped_column(Numeric(16, 3), nullable=True, comment="孔洞总面积 mm²")

    # ── 曲线信息 ──
    curve_length_mm: Mapped[Decimal | None] = mapped_column(Numeric(16, 3), nullable=True, comment="曲线/轮廓长度 mm")
    is_open_curve: Mapped[bool] = mapped_column(Boolean, default=False, comment="是否开放曲线")

    # ── 重叠检测 ──
    overlap_count: Mapped[int | None] = mapped_column(Integer, nullable=True, comment="检测到的重叠对象数")
    overlap_area_mm2: Mapped[Decimal | None] = mapped_column(Numeric(16, 3), nullable=True, comment="重叠区域估计面积 mm²")

    # ── 板材套料 ──
    sheet_count: Mapped[int | None] = mapped_column(Integer, nullable=True, comment="预估需用板材张数")
    sheet_utilization_pct: Mapped[Decimal | None] = mapped_column(Numeric(6, 3), nullable=True, comment="板材利用率（%）")
    sheet_width_mm: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True, comment="使用的板材宽 mm（从 material 快照）")
    sheet_height_mm: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True, comment="使用的板材高 mm")

    # ── 标记 ──
    is_estimated: Mapped[bool] = mapped_column(Boolean, default=True, comment="是否为估算值（非精确测量）")

    # ── 完整数据 ──
    nesting_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True, comment="排版布局结果")
    analysis_json: Mapped[dict | None] = mapped_column(JSONB, nullable=True, comment="完整几何分析明细")
